File: isegoria/src/main/java/com/eulersbridge/isegoria/election/candidates/positions/PositionAdapter.java
package com.eulersbridge.isegoria.election.candidates.positions;

import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bumptech.glide.RequestManager;
import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.network.api.API;
import com.eulersbridge.isegoria.network.api.models.Position;
import com.eulersbridge.isegoria.network.api.responses.PhotosResponse;
import com.eulersbridge.isegoria.util.network.SimpleCallback;
import com.eulersbridge.isegoria.util.ui.LoadingAdapter;

import java.lang.ref.WeakReference;

import retrofit2.Response;

class PositionAdapter extends LoadingAdapter<Position, PositionViewHolder> implements PositionViewHolder.PositionItemListener {

    interface PositionClickListener {
        void onClick(Position item);
    }

    final private PositionClickListener clickListener;
    final private RequestManager glide;
    final private API api;

    PositionAdapter(@NonNull RequestManager glide, API api, @NonNull PositionClickListener clickListener) {
        super(1);

        this.glide = glide;
        this.api = api;
        this.clickListener = clickListener;
    }

    @Override
    public void onViewDetachedFromWindow(PositionViewHolder holder) {
        super.onViewDetachedFromWindow(holder);
    }

    @Override
    public PositionViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        View itemView = LayoutInflater.from(viewGroup.getContext()).
                inflate(R.layout.election_position_grid_item, viewGroup, false);
        return new PositionViewHolder(itemView, this);
    }

    @Override
    public void onClick(Position item) {
        clickListener.onClick(item);
    }

    @Override
    public void getPhoto(PositionViewHolder viewHolder, final long itemId) {
        if (api != null) {

            WeakReference<PositionViewHolder> wrViewHolder = new WeakReference<>(viewHolder);

            api.getPhotos(itemId).enqueue(new SimpleCallback<PhotosResponse>() {
                @Override
                protected void handleResponse(Response<PhotosResponse> response) {
                    PhotosResponse body = response.body();

                    if (body != null && body.totalPhotos > 0 && wrViewHolder.get() != null) {
                        wrViewHolder.get().setImageUrl(glide, body.photos.get(0).thumbnailUrl, itemId);
                    }
                }
            });
        }
    }
}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/election/candidates/positions/PositionAdapter.kt
package com.eulersbridge.isegoria.election.candidates.positions

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.bumptech.glide.RequestManager
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.data.Repository
import com.eulersbridge.isegoria.network.api.model.Position
import com.eulersbridge.isegoria.util.extension.subscribeSuccess
import com.eulersbridge.isegoria.util.ui.LoadingAdapter
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference

internal class PositionAdapter(
    private val glide: RequestManager,
    private val repository: Repository?,
    private val clickListener: PositionClickListener
) :
    LoadingAdapter<Position,
    PositionViewHolder>(1),
    PositionViewHolder.PositionItemListener {

    interface PositionClickListener {
        fun onClick(item: Position)
    }

    private val compositeDisposable = CompositeDisposable()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PositionViewHolder {
        val itemView = LayoutInflater.from(parent.context)
            .inflate(R.layout.election_position_grid_item, parent, false)
        return PositionViewHolder(itemView, this)
    }

    override fun onClick(item: Position?) {
        if (item != null)
            clickListener.onClick(item)
    }

    override fun getPhoto(viewHolder: PositionViewHolder, itemId: Long) {
        if (repository != null) {

            val weakViewHolder = WeakReference(viewHolder)

            repository.getPhotos(itemId).subscribeSuccess {
                it.photos.firstOrNull()?.getPhotoUrl()?.let { photoThumbnailUrl ->
                    weakViewHolder.get()?.apply {
                        setImageUrl(glide, photoThumbnailUrl, itemId)
                    }
                }
            }
        }
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        compositeDisposable.dispose()
    }
}